modules/validations/duplicate_way_segments.js

```diff
@@ -0,0 +1,125 @@
+import { utilDisplayLabel } from '../util';
+import { t } from '../core/localizer';
+import { validationIssue, validationIssueFix } from '../core/validation';
+
+// This validation determines whether way segments are duplicated atop one another,
+// which is impossible to detect via the tool(the geometries are stacked on top of
+// one another, so they are not visible ever).
+export function validationDuplicateWaySegments(context) {
+    var type = 'duplicate_way_segments';
+
+    var validation = function(entity, graph) {
+        if (entity.type === 'way') {
+            return getIssuesForWay(entity);
+        }
+        return [];
+
+        function isRoutableTag(key) {
+            return key === 'highway' ||
+                key === 'railway' ||
+                key === 'waterway';
+        }
+        function isAreaTag(key) {
+            return key === 'area';
+        }
+
+
+    // Consider a way to be routable if it is a highway, railway, or wateray.
+    // if it is an area of any kind, it is not routable.
+    function hasRoutableTags(way) {
+        if (Object.keys(way.tags).some(isAreaTag)) return false;
+        return Object.keys(way.tags).some(isRoutableTag);
+    }
+
+    function adjacentNodes(node1, node2, way) {
+        const nodes = graph.childNodes(way);
+        return Math.abs(nodes.findIndex(node => node.id === node1.id) - nodes.findIndex(node =>  node.id === node2.id)) === 1;
+    }
+
+
+
+        function getIssuesForWay(way) {
+            var issues = [],
```

Comment:
Can we stick the `hasRoutableTags` check here at the very beginning?  I'm thinking it could avoid looping through all the nodes and make this code much more performant..